Add support for parallel_for with 0-sized range

ath0mos opened the pull request:
Added a simple if statement inside of parallel_for which checks the size of numWorkItems

ath0mos committed:
Added if statement to handle the submit of 0 work-items

ath0mos committed:
cleaned up changes made for debugging

ig-raiderler commented:
I'm impressed by the modularity of this codebase. Makes it very maintainable.

ath0mos committed:
Added test-case

ath0mos committed:
Merge branch 'develop' into fix_1764

ath0mos committed:
Included fix also for multidimensional parallel_for-loops

illuhad commented on include/hipSYCL/sycl/handler.hpp:
```diff
@@ -308,63 +308,75 @@ class handler {
+      };
 
-    detail::separate_last_argument_and_apply(invoker, redu_kernel...);
+      detail::separate_last_argument_and_apply(invoker, redu_kernel...);
```
Can you fix the indentation here? We use 2 space indentation everywhere.

illuhad commented on include/hipSYCL/sycl/handler.hpp:
```diff
@@ -308,63 +308,75 @@ class handler {
+      };
 
-    detail::separate_last_argument_and_apply(invoker, redu_kernel...);
+      detail::separate_last_argument_and_apply(invoker, redu_kernel...);
```
Here also

illuhad commented on include/hipSYCL/sycl/handler.hpp:
```diff
@@ -308,63 +308,75 @@ class handler {
                     const ReductionsAndKernel &... redu_kernel) {
-    auto invoker = [&](auto&& kernel, auto&& ... reductions) {
-      this->submit_kernel<KernelName, rt::kernel_type::basic_parallel_for>(
+
```
Why are we not checking `numWorkItems` here as well?

illuhad commented on include/hipSYCL/sycl/handler.hpp:
```diff
@@ -308,63 +308,75 @@ class handler {
                     const ReductionsAndKernel &... redu_kernel) {
-    auto invoker = [&](auto&& kernel, auto&& ... reductions) {
-      this->submit_kernel<KernelName, rt::kernel_type::basic_parallel_for>(
+
```
Why are we not checking `numWorkItems` here as well.

illuhad commented on tests/device_compilation_tests.cpp:
```diff
@@ -171,6 +171,35 @@ struct KernelFunctor {
+  q.parallel_for<>(
+    cl::sycl::range<1>(static_cast<std::size_t>(0)),
+    [=](cl::sycl::id<1> idx) {
+      d_y[idx] = static_cast<int>(idx);
```
Why are we writing something in this test? Was the intention to also add a test to check that the initial data array was not modified by the 0-size kernel? (in that case, data initialization and result comparison would be missing)

illuhad reviewed (commented):
Thanks, some minor remarks :)

ath0mos committed:
Fixed remarks for current commits
   - restored the indentation
   - added if-conditions for loops with workitem-offset
   - changed test-location and added some logic for event-handling

ath0mos committed:
Merge branch 'develop' into fix_1764

ath0mos committed:
Included event & dependency testing also for multi-dim loops

ath0mos committed:
Fixed if check

ath0mos committed:
Fixed minor bugs

ath0mos committed:
Merge branch 'develop' into fix_1764

illuhad commented on tests/sycl/kernel_invocation.cpp:
```diff
@@ -27,6 +27,80 @@ BOOST_AUTO_TEST_CASE(basic_single_task) {
+  auto deps3D = e3D.get_wait_list();
+  BOOST_CHECK(deps1D.empty());
+  BOOST_CHECK(deps2D.empty());
+  BOOST_CHECK(deps3D.empty());
```
Why are we checking command status and `get_wait_list()` here? `wait_and_throw` already sufficiently guarantees that the command has finished.

illuhad commented on include/hipSYCL/sycl/handler.hpp:
```diff
@@ -293,74 +293,96 @@ class handler {
 
   template <typename KernelName = __acpp_unnamed_kernel,
-            typename... ReductionsAndKernel, int dimensions>
+  typename... ReductionsAndKernel, int dimensions>
```
Can we retain the indentation here of `typename...` ? (same for the other functions)

ath0mos committed:
Fixed latest remarks

ath0mos committed:
Merge branch 'develop' into fix_1764

ath0mos committed:
Merge branch 'develop' into fix_1764

ath0mos committed:
Removed unnecessary checks from the test-case

illuhad reviewed (approved):
Thanks! :)
